File: assets/js/theme.test.js
import test from "node:test"
import assert from "node:assert/strict"
import {DARK_THEME, LIGHT_THEME, normalizeTheme, oppositeTheme} from "./theme.js"

test("normalizes supported Tokyo themes", () => {
  assert.equal(normalizeTheme(DARK_THEME), DARK_THEME)
  assert.equal(normalizeTheme(LIGHT_THEME), LIGHT_THEME)
})

test("migrates legacy light and dark preferences", () => {
  assert.equal(normalizeTheme("light"), LIGHT_THEME)
  assert.equal(normalizeTheme("catppuccin-latte"), LIGHT_THEME)
  assert.equal(normalizeTheme("dark"), DARK_THEME)
  assert.equal(normalizeTheme("midnight"), DARK_THEME)
  assert.equal(normalizeTheme("dracula"), DARK_THEME)
})

test("defaults unknown values and toggles within the Tokyo pair", () => {
  assert.equal(normalizeTheme("unknown"), DARK_THEME)
  assert.equal(normalizeTheme(null), DARK_THEME)
  assert.equal(oppositeTheme(DARK_THEME), LIGHT_THEME)
  assert.equal(oppositeTheme(LIGHT_THEME), DARK_THEME)
})

File: test/urielm_web/plugs/theme_test.exs
defmodule UrielmWeb.Plugs.ThemeTest do
  use ExUnit.Case, async: true

  import Plug.Conn
  import Plug.Test

  alias UrielmWeb.Plugs.Theme

  test "defaults to Tokyo Night" do
    conn =
      :get
      |> conn("/")
      |> fetch_cookies()
      |> Theme.call([])

    assert conn.assigns.theme == "tokyo-night"
  end

  test "keeps supported Tokyo preferences" do
    conn = conn_with_theme("tokyo-day")

    assert conn.assigns.theme == "tokyo-day"
  end

  test "migrates legacy light and dark preferences" do
    assert conn_with_theme("light").assigns.theme == "tokyo-day"
    assert conn_with_theme("catppuccin-latte").assigns.theme == "tokyo-day"
    assert conn_with_theme("dark").assigns.theme == "tokyo-night"
    assert conn_with_theme("midnight").assigns.theme == "tokyo-night"
  end

  defp conn_with_theme(theme) do
    :get
    |> conn("/")
    |> put_req_cookie("phx_theme", theme)
    |> fetch_cookies()
    |> Theme.call([])
  end
end
